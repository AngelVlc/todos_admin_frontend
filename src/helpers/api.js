import { requestPending, requestDone, requestFailed } from '../actions'

const backendUrl = process.env.REACT_APP_BACKEND_URL || 'http://localhost:5001'

export const doGetToken = (loginDto, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(loginDto)
        };
        fetch(`${backendUrl}/auth/token`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (!res.ok) {
                    res.text().then(txt => reject(txt));
                    return
                }
                res.json().then(obj => {
                    resolve(obj);
                })
            })
            .catch(error => {
                requestsDispatch(requestDone());
                reject(error.message);
            })
    });
}

const getHeaders = (method, token) => {
    return {
        method: method,
        headers: { 'Content-Type': 'application/json', 'Authorization': `Bearer ${token}` },
    };
}

export const doGet = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('GET', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    res.json().then(obj => {
                        resolve(obj);
                    })
                } else {
                    res.text().then(txt => {
                        requestsDispatch(requestFailed(txt));
                        reject(txt)
                    });
                }
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doDelete = (endpoint, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('DELETE', token);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                if (res.ok) {
                    resolve();
                    return
                }
                res.text().then(txt => {
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}

export const doPost = (endpoint, body, token, requestsDispatch) => {
    return new Promise((resolve, reject) => {
        requestsDispatch(requestPending());
        const requestOptions = getHeaders('POST', token);
        requestOptions.body = JSON.stringify(body);
        fetch(`${backendUrl}/${endpoint}`, requestOptions)
            .then(res => {
                requestsDispatch(requestDone());
                res.text().then(txt => {
                    if (res.ok) {
                        resolve(txt);
                        return
                    }
                    requestsDispatch(requestFailed(txt));
                    reject(txt)
                });
            })
            .catch(error => {
                requestsDispatch(requestFailed(error.message));
                reject(error.message);
            })
    });
}
