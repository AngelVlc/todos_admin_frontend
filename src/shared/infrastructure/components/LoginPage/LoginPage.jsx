import React, { useState, useContext, useEffect } from 'react';
import { AppContext } from '../../contexts';
import { useHistory } from "react-router-dom";
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { userLoggedIn, userLoggedOut } from '../../actions';
import * as Yup from 'yup';
import axios from 'axios';

export const LoginPage = () => {
    const { authDispatch } = useContext(AppContext);
    const [authError, setAuthError] = useState(null);
    let history = useHistory();

    useEffect(() => {
        authDispatch(userLoggedOut());
        localStorage.setItem('userInfo', null);
    }, [authDispatch]);

    const onSubmit = async(values) => {
        try {
            const res = await axios.post('/auth/login', values);
            localStorage.setItem('userInfo', JSON.stringify(res.data));
            authDispatch(userLoggedIn(res.data));
            history.push('/');
        } catch (error) {
            setAuthError(error);
        }
    }

    return (
        <div className="container">
            <h3 className="title">LOG IN</h3>
            <Formik
                initialValues={{
                    name: '',
                    password: ''
                }}
                validationSchema={Yup.object({
                    name: Yup.string().required('Required'),
                    password: Yup.string().required('Required'),
                })}
                onSubmit={onSubmit}>
                <Form>
                    <div className="field">
                        <label className="label" htmlFor="name">Name</label>
                        <div className="control">
                            <Field name="name" type="text" data-testid="name" autoFocus />
                        </div>
                        <p className="help is-danger" data-testid="nameErrors">
                            <ErrorMessage name="name" />
                        </p>
                    </div>

                    <div className="field">
                        <label className="label" htmlFor="password">Password</label>
                        <div className="control">
                            <Field name="password" type="password" data-testid="password" />
                        </div>
                        <p className="help is-danger" data-testid="passwordErrors">
                            <ErrorMessage name="password" />
                        </p>
                    </div>
                    <div className="control">
                        <button className="button" type="submit" data-testid="submit">Log In</button>
                        {authError ? (
                            <p className="help is-danger" data-testid="authError">{authError}</p>
                        ) : null}
                    </div>
                </Form>
            </Formik>
        </div>
    );
}
