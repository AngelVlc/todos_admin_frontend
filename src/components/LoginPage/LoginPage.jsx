import React, { useContext, useState } from 'react';
import { useHistory } from "react-router-dom";
import { useFormik } from 'formik';
import { userLoggedIn } from '../../actions'
import { AppContext } from '../../contexts/AppContext'
import * as Yup from 'yup';
import jwt from 'jwt-decode';
import { doGetToken } from '../../helpers/api';


export const LoginPage = () => {
    const { authDispatch, requestsDispatch } = useContext(AppContext)
    const [authError, setAuthError] = useState(null);
    let history = useHistory();

    const formik = useFormik({
        initialValues: {
            userName: '',
            password: ''
        },
        validationSchema: Yup.object({
            userName: Yup.string()
                .required('Required'),
            password: Yup.string()
                .required('Required'),
        }),
        onSubmit: async (values) => {
            try {
                const res = await doGetToken(values, requestsDispatch);
                const tokenInfo = jwt(res.token);
                authDispatch(userLoggedIn(res, tokenInfo));
                history.push('/');
            } catch (error) {
                setAuthError(error);
            }
        },
    });

    return (
        <div>
            <h4>LOG IN PAGE</h4>
            <form onSubmit={formik.handleSubmit}>
                <label htmlFor="userName">Name</label>
                <input
                    id="userName"
                    name="userName"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    value={formik.values.userName}
                />
                {formik.touched.userName && formik.errors.userName ? (
                    <div>{formik.errors.userName}</div>
                ) : null}
                <label htmlFor="password">Password</label>
                <input
                    id="password"
                    name="password"
                    type="password"
                    onChange={formik.handleChange}
                    onBlur={formik.handleBlur}
                    value={formik.values.password}
                />
                {formik.touched.password && formik.errors.password ? (
                    <div>{formik.errors.password}</div>
                ) : null}
                {authError ? (
                    <div>{authError}</div>
                ) : null}
                <button type="submit">Log In</button>
            </form>
        </div>
    );
}
