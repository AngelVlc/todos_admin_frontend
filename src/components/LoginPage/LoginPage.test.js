import React from 'react'
import { render, cleanup, fireEvent, wait } from '@testing-library/react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import { LoginPage } from './LoginPage'
import { AppContext } from '../../contexts/AppContext'
import * as api from '../../helpers/api';
import jwt from 'jwt-decode';

jest.mock('jwt-decode');
const mockAuthDispatch = jest.fn()

const renderWithContextAndRouter = (component) => {
    const history = createMemoryHistory()
    const context = { authDispatch: mockAuthDispatch }
    return {
        ...render(
            <AppContext.Provider value={context}>
                <Router history={history}>
                    {component}
                </Router>
            </AppContext.Provider>)
    }
}

afterEach(cleanup)

it('should take a snapshot', () => {
    const { asFragment } = renderWithContextAndRouter(<LoginPage />);

    expect(asFragment(<LoginPage />)).toMatchSnapshot();
});

it('should require user name and password for logging in', async () => {
    const { getByTestId } = renderWithContextAndRouter(<LoginPage />);

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(getByTestId('userNameErrors')).toHaveTextContent('Required');
    expect(getByTestId('passwordErrors')).toHaveTextContent('Required');
})

it('should show the error when log in fails', async () => {
    const { getByTestId } = renderWithContextAndRouter(<LoginPage />);

    await wait(() => {
        fireEvent.change(getByTestId('userName'), {
            target: {
                value: "user"
            }
        })
    })

    await wait(() => {
        fireEvent.change(getByTestId('password'), {
            target: {
                value: "pass"
            }
        })
    })

    api.doGetToken = jest.fn((loginDto, requestDispatch) => {
        return Promise.reject('some error');
    });

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(getByTestId('authError')).toHaveTextContent('some error');
})


it('should show the home page after logging in', async () => {
    const { container, getByTestId } = renderWithContextAndRouter(<LoginPage />);

    await wait(() => {
        fireEvent.change(getByTestId('userName'), {
            target: {
                value: 'user'
            }
        })
    })

    await wait(() => {
        fireEvent.change(getByTestId('password'), {
            target: {
                value: 'pass'
            }
        })
    })

    api.doGetToken = jest.fn((loginDto, requestDispatch) => {
        const tokens = {
            'refreshToken': 'theRefreshToken',
            'token': 'theToken'
        }
        return Promise.resolve(tokens);
    });

    jwt.mockImplementation(() => {
        return {
            exp: 'exp',
            userId: 10,
            userName: 'user'
        }
    })

    await wait(() => {
        fireEvent.click(getByTestId('submit'));
    })

    expect(mockAuthDispatch.mock.calls.length).toBe(1);
    const userLoggedIn = {
        type: 'USER_LOGGED_IN',
        authInfo: {
            exp: 'exp',
            token: 'theToken',
            refreshToken: 'theRefreshToken',
            userName: 'user',
            userId: 10
        }
    }
    expect(mockAuthDispatch.mock.calls[0][0]).toStrictEqual(userLoggedIn);
    expect(location.pathname).toBe("/");
})
