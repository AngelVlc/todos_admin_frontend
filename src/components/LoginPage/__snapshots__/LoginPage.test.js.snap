// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4>
      LOG IN PAGE
    </h4>
    <form>
      <label
        for="userName"
      >
        Name
      </label>
      <input
        data-testid="userName"
        id="userName"
        name="userName"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        data-testid="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        data-testid="submit"
        type="submit"
      >
        Log In
      </button>
    </form>
  </div>
</DocumentFragment>
`;
